cdap-unit-test/src/test/java/co/cask/cdap/partitioned/concatenate/PartitionConcatenateTest.java

```diff
@@ -0,0 +1,143 @@
+ * Tests functionality to concatenate files within a partition of a PartitionedFileSet.
+ */
+public class PartitionConcatenateTest extends TestFrameworkTestBase {
+
+  /**
+   * 1. Write 100 small files (text) to an input FileSet.
+   * 2. Run a MapReduce that reads these files and writes them in ORC format to a Partition of a PartitionedFileSet.
+   * 3. Execute a partition concatenate operation.
+   * 4. As compared to before the concatenate operation, validate that the number of files is reduced, while
+   *    the contents of the files remains the same.
+   */
+  @Test
+  public void testConcatenate() throws Exception {
+    ApplicationManager appManager = deployApplication(OrcApp.class);
+
+    // 1. create 100 small files in the input FileSet
+    int numInputFiles = 100;
+    DataSetManager<FileSet> inputFileSet = getDataset(OrcApp.FILESET_NAME);
+    List<String> writtenData = writeSmallFiles(inputFileSet.get().getBaseLocation(), numInputFiles);
+
+    // 2. run the MapReduce job to write these 100 files to a PartitionedFileSet, with orc format
+    long outputPartitionTime = 5000;
+    MapReduceManager mapReduceManager = appManager.getMapReduceManager(OrcApp.MyMapReduce.NAME)
+      .start(Collections.singletonMap(OrcApp.MyMapReduce.OUTPUT_PARTITION_KEY, Long.toString(outputPartitionTime)));
+    mapReduceManager.waitForRuns(ProgramRunStatus.COMPLETED, 1, 30, TimeUnit.SECONDS);
+
+    Assert.assertEquals(writtenData, getExploreResults());
+
+    DataSetManager<PartitionedFileSet> cleanRecordsManager = getDataset(OrcApp.CLEAN_RECORDS);
+    PartitionedFileSet cleanRecords = cleanRecordsManager.get();
+
+    PartitionKey outputPartition = PartitionKey.builder().addLongField("time", outputPartitionTime).build();
+    Location partitionLocation = cleanRecords.getPartition(outputPartition).getLocation();
+
+    // this is a timestamp before concatenating, but after running the MapReduce job
+    long beforeConcatTime = System.currentTimeMillis();
+
+    List<Location> dataFiles = listFilteredChildren(partitionLocation);
+    // each input file will result in one output file, due to the FileInputFormat class and FileOutputFormat class
+    // being used
+    Assert.assertEquals(numInputFiles, dataFiles.size());
+    for (Location dataFile : dataFiles) {
+      // all the files should have a lastModified smaller than now
+      Assert.assertTrue(dataFile.lastModified() < beforeConcatTime);
+    }
+
+    // 3. run the concatenate operation
+    cleanRecords.concatenatePartition(outputPartition);
+
+    // 4. check that the data files' lastModified timestamp is updated, and there should be fewer of them
+    dataFiles = listFilteredChildren(partitionLocation);
+    int fileCount = dataFiles.size();
+    Assert.assertTrue(fileCount < numInputFiles);
+    // note that Hive doesn't guarantee the number of output files of a CONCATENATE operation, so the
+    // following assertion (fileCount == 1) may not always be true, but from observation, it is
```

Comment:
this doesn't sound very assuring. Maybe we can check there are fewer than 100 files and content is the same?